google/cloud/pubsub/samples/samples.cc

```diff
@@ -518,27 +518,29 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
   using google::cloud::future;
   using google::cloud::StatusOr;
   [](pubsub::Subscriber subscriber, pubsub::Subscription const& subscription) {
-    std::atomic<int> count{0};
-    google::cloud::promise<void> received_message;
+    std::mutex mu;
+    std::condition_variable cv;
+    bool done = false;
+    int message_count = 0;
     auto result = subscriber.Subscribe(
         subscription, [&](pubsub::Message const& m, pubsub::AckHandler h) {
           std::cout << "Received message " << m << "\n";
-          // Signal the waiting thread to detach the subscription after the
-          // first message.
-          if (count.load() == 0) received_message.set_value();
-          ++count;
-          // Only then ack the message, preventing any new requests from being
-          // issued.
+          std::unique_lock<std::mutex> lk(mu);
+          ++message_count;
+          done = true;
+          cv.notify_one();
+          lk.unlock();
           std::move(h).ack();
         });
-    // Cancel the subscription as soon as the first message is received, and
-    // block until that happens.
-    received_message.get_future()
-        .then([&result](google::cloud::future<void>) { result.cancel(); })
-        .get();
+    // Wait until at least one message has been received.
+    std::unique_lock<std::mutex> lk(mu);
+    cv.wait(lk, [&done] { return done; });
+    lk.unlock();
+    // Cancel the subscription session.
+    result.cancel();
     // Report any final status, blocking.
-    std::cout << "Message count = " << count.load()
-              << ", status = " << result.get() << "\n";
+    std::cout << "Message count: " << message_count
```

Comment:
Opt: If we were using lock annotations, we'd probably like to say that "`message_count` is guarded by `mu`", and that would generate an "unsynchronized access" warning here, methinks.  So consider whether it might be clearer to snapshot `message_count` under the lock even if the `cancel()` has guaranteed that we're "after" all the subscription callbacks.  Or, perhaps just a comment to that effect.

Reply:
I went with the comment and a small code change.